proposal/first-class-calc.md

```diff
@@ -0,0 +1,541 @@
+  Otherwise, emit `left`.
+
+* Emit `" "`, then the operator, then `" "`.
+
+* If the operator is `"*"` or `"/"` and the right value is a `CalcOperation`
+  with operator `"+"` or `"-"`, emit `"("` followed by `right` followed by
+  `")"`. Otherwise, emit `right`.
+
+  > TODO: If one of the operands is a result of an interpolated expression, it
+  > may need parentheses. However, we *don't* want to add parentheses for
+  > unquoted strings from e.g. `var()` expressions. We need to find a way to
+  > distinguish the two.
+
+## Procedures
+
+### Simplifying a Calc
+
+This algorithm takes a calc `calc` and returns a number or a calc.
+
+> This algorithm is intended to return a value that's CSS-semantically identical
+> to the input.
+
+* Let `arguments` be the result of [simplifying](#simplifying-a-calcvalue) each
+  of `calc`'s arguments.
+
+* If `calc`'s name is `"calc"`, the syntax guarantees that `arguments` contain
+  only a single argument. If that argument is a number or calc, return it.
+
+* If `calc`'s name is `min`, `max`, or `clamp` and `arguments` are all numbers
+  whose units are mutually [compatible], return the result of calling
+  [`math.min()`], [`math.max()`], or `math.clamp()` (respectively) with those
+  arguments.
+
+  [compatible]: ../spec/types/number.md#compatible-units
+  [`math.min()`]: ../spec/built_in_modules/math.md#min
+  [`math.max()`]: ../spec/built_in_modules/math.md#max
+
+* Otherwise, return a calc with the same name as `calc` and `arguments` as its
+  arguments.
+
+### Simplifying a `CalcValue`
+
+This algorithm takes a `CalcValue` `value` and returns a `CalcValue`.
+
+> This algorithm is intended to return a value that's CSS-semantically identical
+> to the input.
+
+* If `value` is a number or unquoted string, return it as-is.
+
+* If `value` is a calc:
+
+    * Let `result` be the result of [simplifying] `value`.
+
+    * If `result` is a calc whose name is `"calc"`, return `result`'s single
+      argument.
+
+      > TODO: If `result` was created via interpolation, it may not be
+      > syntactically sound to just strip the `calc()`. For example, `calc(-1 *
+      > calc(#{"1px + 10%"})` should return `calc(-1 * (1px + 10%))` rather than
+      > `calc(-1 * 1px + 10%)`.
+
+    * Otherwise, return `result`.
+
+  [simplifying]: #simplifying-a-calc
+
+* Otherwise, `value` must be a `CalcOperation`. Let `left` and `right` be the
+  result of simplifying `value.left` and `value.right`, respectively.
+
+* If `value.operator` is `"+"` or `"-"`:
+
+  * If `left` and `right` are both numbers with [compatible] units, return
+    `left + right` or `left - right`, respectively.
+
+    > TODO: Should we throw an error here for units we can prove are actively
+    > incompatible? For example, unitless numbers are always incompatible with
+    > numbers of any units, and numbers across different known types (length +
+    > time) are also invalid.
+
+    > TODO: Should we try to simplify `calc(1px + 1% + 1px)`?
```

Comment:
For here and other similar TODOs below: what are the cons of simplifying this as much as possible? It seems to me that simplifying to `calc(2px + 1%)` would be the most logically consistent.

Reply:
This would require a more complex structure for the AST, because the `1px + 1% + 1px` would be represented as `CalcOperation(1px, '+', CalcOperation(1%, '+', 1px))` currently while simplifying such case would require recombining the sum to try another variant.

Reply:
Alternately, it would require complexity in the simplification logic to whether `CalcOperation`s can be "flattened" usefully. The CSS spec actually does cover this sort of simplification logic, so there's prior art, but it's definitely more complex.